README.rst
talkiq/yaaredis
===============

|circleci| |pypi-version| |python-versions|

.. |circleci| image:: https://img.shields.io/circleci/project/github/talkiq/yaaredis/master.svg?style=flat-square
    :alt: CircleCI Test Status
    :target: https://circleci.com/gh/talkiq/yaaredis/tree/master

.. |pypi-version| image:: https://img.shields.io/pypi/v/yaaredis.svg?style=flat-square&label=PyPI
    :alt: Latest PyPI Release
    :target: https://pypi.org/project/yaaredis/

.. |python-versions| image:: https://img.shields.io/pypi/pyversions/yaaredis.svg?style=flat-square&label=Python%20Versions
    :alt: Compatible Python Versions
    :target: https://pypi.org/project/yaaredis/

``yaaredis`` (Yet Another Async Redis (client)) is a fork of
`aredis <https://github.com/NoneGG/aredis>`_, which itself was ported from
`redis-py <https://github.com/andymccurdy/redis-py>`_. ``yaaredis`` provides an
efficient and user-friendly async redis client with support for Redis Server,
Cluster, and Sentinels.

To get more information please read the `full documentation`_ managed by the
upstream ``aredis`` repo. We are working on hosting our own as the projects
diverge -- stay tuned!

Installation
------------

``yaaredis`` requires a running Redis server. To install yaaredis, simply:

.. code-block:: console

    python3 -m pip install yaaredis

or from source:

.. code-block:: console

    python3 -m pip install .

Note that ``yaaredis`` also supports using ``hiredis`` as a drop-in performance
improvements. You can either install ``hiredis`` separately or make use of the
PyPI extra to make use of this functionality:

.. code-block:: console

    python3 -m pip install yaaredis[hiredis]

Getting started
---------------

We have `various examples`_ in this repo which you may find useful. A few more
specific cases are listed below.

Single Node Client
^^^^^^^^^^^^^^^^^^

.. code-block:: python

    import asyncio
    from yaaredis import StrictRedis

    async def example():
        client = StrictRedis(host='127.0.0.1', port=6379, db=0)
        await client.flushdb()
        await client.set('foo', 1)
        assert await client.exists('foo') is True
        await client.incr('foo', 100)

        assert int(await client.get('foo')) == 101
        await client.expire('foo', 1)
        await asyncio.sleep(0.1)
        await client.ttl('foo')
        await asyncio.sleep(1)
        assert not await client.exists('foo')

    asyncio.run(example())

Cluster Client
^^^^^^^^^^^^^^

.. code-block:: python

    import asyncio
    from yaaredis import StrictRedisCluster

    async def example():
        client = StrictRedisCluster(host='172.17.0.2', port=7001)
        await client.flushdb()
        await client.set('foo', 1)
        await client.lpush('a', 1)
        print(await client.cluster_slots())

        await client.rpoplpush('a', 'b')
        assert await client.rpop('b') == b'1'

    asyncio.run(example())
    # {(10923, 16383): [{'host': b'172.17.0.2', 'node_id': b'332f41962b33fa44bbc5e88f205e71276a9d64f4', 'server_type': 'master', 'port': 7002},
    # {'host': b'172.17.0.2', 'node_id': b'c02deb8726cdd412d956f0b9464a88812ef34f03', 'server_type': 'slave', 'port': 7005}],
    # (5461, 10922): [{'host': b'172.17.0.2', 'node_id': b'3d1b020fc46bf7cb2ffc36e10e7d7befca7c5533', 'server_type': 'master', 'port': 7001},
    # {'host': b'172.17.0.2', 'node_id': b'aac4799b65ff35d8dd2ad152a5515d15c0dc8ab7', 'server_type': 'slave', 'port': 7004}],
    # (0, 5460): [{'host': b'172.17.0.2', 'node_id': b'0932215036dc0d908cf662fdfca4d3614f221b01', 'server_type': 'master', 'port': 7000},
    # {'host': b'172.17.0.2', 'node_id': b'f6603ab4cb77e672de23a6361ec165f3a1a2bb42', 'server_type': 'slave', 'port': 7003}]}

Benchmark
---------

Please run test scripts in the ``benchmarks`` directory to confirm the
benchmarks. For a benchmark in the original yaaredis author's environment
please see: `benchmark`_.

Contributing
------------

Developer? See our `guide`_ on how you can contribute.

.. _benchmark: http://aredis.readthedocs.io/en/latest/benchmark.html
.. _full documentation: http://aredis.readthedocs.io/en/latest/
.. _guide: https://github.com/talkiq/yaaredis/blob/master/.github/CONTRIBUTING.rst
.. _various examples: https://github.com/talkiq/yaaredis/tree/master/examples


CHANGELOG.md
<a name="3.0.0"></a>
## 3.0.0 (2022-01-17)

#### Features

* **compat:**  add support for python 3.10 (#13) ([dda3b5de](dda3b5de))

#### Bug Fixes

* **pool:**  fix basic query argument types ([30de2852](30de2852))

#### Breaking Changes

* **compat:**  drop support for python 3.5 (#13) ([dda3b5de](dda3b5de))

<a name="2.0.4"></a>
## 2.0.4 (2021-07-19)

#### Bug Fixes

* **connection:**  avoid hiding timeout errors on connect (#11) ([7186a9b4](7186a9b4))

<a name="2.0.3"></a>
## 2.0.3 (2021-07-02)

#### Bug Fixes

* **cluster:**  fix broken retry logic for ClusterDownErrors (#8) ([a16c35d2](a16c35d2))

<a name="2.0.2"></a>
## 2.0.2 (2021-06-23)

#### Bug Fixes

* **cluster**:  fixup invalid `Cluster.scan_iter()` default value for `type` param (#10) ([2064e842](2064e842))
* **cluster**:  fixup support for Python 3.6 in `Cluster.scan_iter()` (#10) ([2064e842](2064e842))
* **cluster**:  avoid potential pagination race condition in `Cluster.scan_iter()` (#10) ([2064e842](2064e842))

<a name="2.0.1"></a>
## 2.0.1 (2021-06-18)

#### Performance

* **cluster:**  run scan operation in parallel across nodes (#7) ([4fbb46d0](4fbb46d0))

<a name="2.0.0"></a>
## 2.0.0 (2021-06-08)

This is a breaking change! As of v1.1.8+, `aredis` has been hard-forked and
renamed to `yaaredis`. The `aredis` changelog is maintained below (see
`1.1.8`_ for posterity).

#### Breaking Changes

* rename package to `yaaredis` ([1ea13dd](1ea13dd))
* remove all deprecated features ([f4699a45](f4699a45)):
  * deprecated methods have been removed
  * deprecated options now raise Exceptions
  * some ignorable deprecations have been replaced with error logs
* **pipeline:**  auto-execute all remaining commands at end of `with` block ([814ca2bb](814ca2bb))
* **pool**:  `ConnectionPool.get_connection()` is now a coroutine ([32969ed9](32969ed9))

#### Features

* **auth:**  add support for Redis 5+ user&pass auth ([ca96dc6d](ca96dc6d))
* **client:**  add option to init with client name ([f00c361a](f00c361a))
* **commands:**  add support for SET's KEEPTTL option ([22c62f04](22c62f04))
* **pool:**  implement blocking connection pool ([32969ed9](32969ed9)):
  * this is available as the `BlockingConnectionPool` class; note that it does
    not yet support cluster mode
* **scan:**  add TYPE option ([0625dd8c](0625dd8c))

#### Bug Fixes

* **client:**  prevent retry_on_timeout from affecting ConnectionError ([6626ebce](6626ebce)):
  * ie. `retry_on_timeout` now only comes into play for timeouts
* **cluster:**  avoid runtime error on disconnected client ([740ea19f](740ea19f)):
  * prevents a race condition within `yaaredis` from raising occasional errors
* **connection:**
  *  always reduce count on force disconnect ([b49b3c65](b49b3c65)):
    * fixes some instances of "connection leaking"
  *  expose initial `connect()` error message ([2073f576](2073f576)):
    * by forwarding this message, this should make some debugging a bit easier
* **pubsub:**  reraise CancelledError during _execute ([90dd2641](90dd2641)):
  * prevents an issue where a cancelled coroutine may continue indefinitely

#### Performance

* **cluster:**  support MGET and MSET for hashed keys ([08180835](08180835)):
  * see the following redis documentation for more info:
    https://redis.io/topics/cluster-tutorial#redis-cluster-data-sharding
    you'll need to make use of "hash slots" to get the full benefit of this
    feature

#### Internal

* convert build system to poetry ([b86d695](b86d695)):
  * note that this should not affect end-users, only yaaredis developers
* publish pypi wheels from CI ([05c2265](05c2265)):
  * we should now have wheels for various versions of Python for all releases
  * we currently support manylinux wheels for all compatible Python versions

<a name="1.1.8"></a>
## 1.1.8
* Fixbug: connection is disconnected before idel check, valueError will be raised if a connection(not exist) is removed from connection list
* Fixbug: abstract compat.py to handle import problem of asyncio.future
* Fixbug: When cancelling a task, CancelledError exception is not propagated to client
* Fixbug: XREAD command should accept 0 as a block argument
* Fixbug: In redis cluster mode, XREAD command does not function properly
* Fixbug: slave connection params when there are no slaves

<a name="1.1.7"></a>
## 1.1.7
* Fixbug: ModuleNotFoundError raised when install aredis 1.1.6 with Python3.6

<a name="1.1.6"></a>
## 1.1.6
* Fixbug: parsing stream messgae with empty payload will cause error(#116)
* Fixbug: Let ClusterConnectionPool handle skip_full_coverage_check (#118)

aredis/client.py
import asyncio
import sys

from aredis.commands.cluster import ClusterCommandMixin
from aredis.commands.connection import ClusterConnectionCommandMixin, ConnectionCommandMixin
from aredis.commands.extra import ExtraCommandMixin
from aredis.commands.geo import GeoCommandMixin
from aredis.commands.hash import ClusterHashCommandMixin, HashCommandMixin
from aredis.commands.hyperlog import ClusterHyperLogCommandMixin, HyperLogCommandMixin
from aredis.commands.keys import ClusterKeysCommandMixin, KeysCommandMixin
from aredis.commands.lists import ClusterListsCommandMixin, ListsCommandMixin
from aredis.commands.pubsub import CLusterPubSubCommandMixin, PubSubCommandMixin
from aredis.commands.scripting import ClusterScriptingCommandMixin, ScriptingCommandMixin
from aredis.commands.sentinel import ClusterSentinelCommands, SentinelCommandMixin
from aredis.commands.server import ClusterServerCommandMixin, ServerCommandMixin
from aredis.commands.sets import ClusterSetsCommandMixin, SetsCommandMixin
from aredis.commands.sorted_set import ClusterSortedSetCommandMixin, SortedSetCommandMixin
from aredis.commands.streams import StreamsCommandMixin
from aredis.commands.strings import ClusterStringsCommandMixin, StringsCommandMixin
from aredis.commands.transaction import ClusterTransactionCommandMixin, TransactionCommandMixin
from aredis.compat import CancelledError
from aredis.connection import RedisSSLContext, UnixDomainSocketConnection
from aredis.exceptions import (AskError, BusyLoadingError, ClusterDownError, ClusterError, ConnectionError, MovedError,
                               RedisClusterException, TimeoutError, TryAgainError)
from aredis.pool import (ClusterConnectionPool, ConnectionPool)
from aredis.utils import (NodeFlag, blocked_command, clusterdown_wrapper, dict_merge, first_key)

mixins = [
    ClusterCommandMixin, ConnectionCommandMixin, ExtraCommandMixin,
    GeoCommandMixin, HashCommandMixin, HyperLogCommandMixin,
    KeysCommandMixin, ListsCommandMixin, PubSubCommandMixin,
    ScriptingCommandMixin, SentinelCommandMixin, ServerCommandMixin,
    SetsCommandMixin, SortedSetCommandMixin, StringsCommandMixin,
    TransactionCommandMixin, StreamsCommandMixin
]

cluster_mixins = [
    ClusterCommandMixin, ClusterStringsCommandMixin, ClusterServerCommandMixin,
    ClusterConnectionCommandMixin, CLusterPubSubCommandMixin, ClusterSentinelCommands,
    ClusterKeysCommandMixin, ClusterScriptingCommandMixin, ClusterHashCommandMixin,
    ClusterSetsCommandMixin, ClusterSortedSetCommandMixin, ClusterTransactionCommandMixin,
    ClusterListsCommandMixin, ClusterHyperLogCommandMixin
]

if sys.version_info[:2] >= (3, 6):
    from aredis.commands.iter import IterCommandMixin, ClusterIterCommandMixin

    mixins.append(IterCommandMixin)
    cluster_mixins.append(ClusterIterCommandMixin)


class StrictRedis(*mixins):
    """
    Implementation of the Redis protocol.

    This abstract class provides a Python interface to all Redis commands
    and an implementation of the Redis protocol.

    Connection and Pipeline derive from this, implementing how
    the commands are sent and received to the Redis server
    """

    RESPONSE_CALLBACKS = dict_merge(*[mixin.RESPONSE_CALLBACKS for mixin in mixins])

    @classmethod
    def from_url(cls, url, db=None, **kwargs):
        """
        Return a Redis client object configured from the given URL, which must
        use either `the ``redis://`` scheme
        <http://www.iana.org/assignments/uri-schemes/prov/redis>`_ for RESP
        connections or the ``unix://`` scheme for Unix domain sockets.

        For example:

        redis://[:password]@localhost:6379/0
        unix://[:password]@/path/to/socket.sock?db=0

        There are several ways to specify a database number. The parse function
        will return the first specified option:
        1. A ``db`` querystring option, e.g. redis://localhost?db=0
        2. If using the redis:// scheme, the path argument of the url, e.g.
        redis://localhost/0
        3. The ``db`` argument to this function.

        If none of these options are specified, db=0 is used.

        Any additional querystring arguments and keyword arguments will be
        passed along to the ConnectionPool class's initializer. In the case
        of conflicting arguments, querystring arguments always win.
        """
        connection_pool = ConnectionPool.from_url(url, db=db, **kwargs)
        return cls(connection_pool=connection_pool)

    def __init__(self, host='localhost', port=6379,
                 db=0, username=None, password=None, stream_timeout=None,
                 connect_timeout=None, connection_pool=None,
                 unix_socket_path=None, encoding='utf-8',
                 decode_responses=False, ssl=False, ssl_context=None,
                 ssl_keyfile=None, ssl_certfile=None,
                 ssl_cert_reqs=None, ssl_ca_certs=None,
                 max_connections=None, retry_on_timeout=False,
                 max_idle_time=0, idle_check_interval=1,
                 client_name=None, loop=None, **kwargs):
        if not connection_pool:
            kwargs = {
                'db': db,
                'username': username,
                'password': password,
                'encoding': encoding,
                'stream_timeout': stream_timeout,
                'connect_timeout': connect_timeout,
                'max_connections': max_connections,
                'retry_on_timeout': retry_on_timeout,
                'decode_responses': decode_responses,
                'max_idle_time': max_idle_time,
                'idle_check_interval': idle_check_interval,
                'client_name': client_name,
                'loop': loop
            }
            # based on input, setup appropriate connection args
            if unix_socket_path is not None:
                kwargs.update({
                    'path': unix_socket_path,
                    'connection_class': UnixDomainSocketConnection
                })
            else:
                # TCP specific options
                kwargs.update({
                    'host': host,
                    'port': port
                })
                if ssl_context is not None:
                    kwargs['ssl_context'] = ssl_context
                elif ssl:
                    ssl_context = RedisSSLContext(ssl_keyfile, ssl_certfile, ssl_cert_reqs, ssl_ca_certs).get()
                    kwargs['ssl_context'] = ssl_context
            connection_pool = ConnectionPool(**kwargs)
        self.connection_pool = connection_pool
        self._use_lua_lock = None

        self.response_callbacks = self.__class__.RESPONSE_CALLBACKS.copy()

    def __repr__(self):
        return "{}<{}>".format(type(self).__name__, repr(self.connection_pool))

    def set_response_callback(self, command, callback):
        """Sets a custom Response Callback"""
        self.response_callbacks[command] = callback

    # COMMAND EXECUTION AND PROTOCOL PARSING
    async def execute_command(self, *args, **options):
        """Executes a command and returns a parsed response"""
        pool = self.connection_pool
        command_name = args[0]
        connection = await pool.get_connection()
        try:
            await connection.send_command(*args)
            return await self.parse_response(connection, command_name, **options)
        except CancelledError:
            # do not retry when coroutine is cancelled
            connection.disconnect()
            raise
        except ConnectionError:
            connection.disconnect()
            raise
        except TimeoutError:
            connection.disconnect()
            if not connection.retry_on_timeout:
                raise

            await connection.send_command(*args)
            return await self.parse_response(connection, command_name, **options)
        finally:
            pool.release(connection)

    async def parse_response(self, connection, command_name, **options):
        """Parses a response from the Redis server"""
        response = await connection.read_response()
        if command_name in self.response_callbacks:
            callback = self.response_callbacks[command_name]
            return callback(response, **options)
        return response

    async def pipeline(self, transaction=True, shard_hint=None):
        """
        Returns a new pipeline object that can queue multiple commands for
        later execution. ``transaction`` indicates whether all commands
        should be executed atomically. Apart from making a group of operations
        atomic, pipelines are useful for reducing the back-and-forth overhead
        between the client and server.
        """
        from aredis.pipeline import StrictPipeline
        pipeline = StrictPipeline(self.connection_pool, self.response_callbacks,
                                  transaction, shard_hint)
        await pipeline.reset()
        return pipeline


class StrictRedisCluster(StrictRedis, *cluster_mixins):
    """
    If a command is implemented over the one in StrictRedis then it requires some changes compared to
    the regular implementation of the method.
    """
    RedisClusterRequestTTL = 16
    NODES_FLAGS = dict_merge(*[mixin.NODES_FLAGS
                               for mixin in cluster_mixins
                               if hasattr(mixin, 'NODES_FLAGS')])
    RESULT_CALLBACKS = dict_merge(*[mixin.RESULT_CALLBACKS
                                    for mixin in cluster_mixins
                                    if hasattr(mixin, 'RESULT_CALLBACKS')])

    def __init__(self, host=None, port=None, startup_nodes=None, max_connections=32,
                 max_connections_per_node=False, readonly=False,
                 reinitialize_steps=None, skip_full_coverage_check=False,
                 nodemanager_follow_cluster=False, **kwargs):
        """
        :startup_nodes:
        List of nodes that initial bootstrapping can be done from
        :host:
        Can be used to point to a startup node
        :port:
        Can be used to point to a startup node
        :max_connections:
        Maximum number of connections that should be kept open at one time
        :readonly:
        enable READONLY mode. You can read possibly stale data from slave.
        :skip_full_coverage_check:
        Skips the check of cluster-require-full-coverage config, useful for clusters
        without the CONFIG command (like aws)
        :nodemanager_follow_cluster:
        The node manager will during initialization try the last set of nodes that
        it was operating on. This will allow the client to drift along side the cluster
        if the cluster nodes move around alot.
        :**kwargs:
        Extra arguments that will be sent into StrictRedis instance when created
        (See Official redis-py doc for supported kwargs
        [https://github.com/andymccurdy/redis-py/blob/master/redis/client.py])
        Some kwargs is not supported and will raise RedisClusterException
        - db (Redis do not support database SELECT in cluster mode)
        """
        # Tweaks to StrictRedis client arguments when running in cluster mode
        if "db" in kwargs:
            raise RedisClusterException("Argument 'db' is not possible to use in cluster mode")
        if "connection_pool" in kwargs:
            pool = kwargs.pop('connection_pool')
        else:
            startup_nodes = [] if startup_nodes is None else startup_nodes

            # Support host/port as argument
            if host:
                startup_nodes.append({"host": host, "port": port if port else 7000})
            pool = ClusterConnectionPool(
                startup_nodes=startup_nodes,
                max_connections=max_connections,
                reinitialize_steps=reinitialize_steps,
                max_connections_per_node=max_connections_per_node,
                skip_full_coverage_check=skip_full_coverage_check,
                nodemanager_follow_cluster=nodemanager_follow_cluster,
                readonly=readonly,
                **kwargs
            )

        super(StrictRedisCluster, self).__init__(connection_pool=pool, **kwargs)

        self.refresh_table_asap = False
        self.nodes_flags = self.__class__.NODES_FLAGS.copy()
        self.result_callbacks = self.__class__.RESULT_CALLBACKS.copy()
        self.response_callbacks = self.__class__.RESPONSE_CALLBACKS.copy()

    @classmethod
    def from_url(cls, url, db=None, skip_full_coverage_check=False, **kwargs):
        """
        Return a Redis client object configured from the given URL, which must
        use either `the ``redis://`` scheme
        <http://www.iana.org/assignments/uri-schemes/prov/redis>`_ for RESP
        connections or the ``unix://`` scheme for Unix domain sockets.
        For example::
        redis://[:password]@localhost:6379/0
        unix://[:password]@/path/to/socket.sock?db=0
        There are several ways to specify a database number. The parse function
        will return the first specified option:
        1. A ``db`` querystring option, e.g. redis://localhost?db=0
        2. If using the redis:// scheme, the path argument of the url, e.g.
        redis://localhost/0
        3. The ``db`` argument to this function.
        If none of these options are specified, db=0 is used.
        Any additional querystring arguments and keyword arguments will be
        passed along to the ConnectionPool class's initializer. In the case
        of conflicting arguments, querystring arguments always win.
        """
        connection_pool = ClusterConnectionPool.from_url(url, db=db, skip_full_coverage_check=skip_full_coverage_check,
                                                         **kwargs)
        return cls(connection_pool=connection_pool)

    def __repr__(self):
        servers = list({'{0}:{1}'.format(info['host'], info['port'])
                        for info in self.connection_pool.nodes.startup_nodes})
        servers.sort()
        return "{0}<{1}>".format(type(self).__name__, ', '.join(servers))

    def set_result_callback(self, command, callback):
        "Sets a custom Result Callback"
        self.result_callbacks[command] = callback

    def _determine_slot(self, *args):
        """Figures out what slot based on command and args"""
        if len(args) <= 1:
            raise RedisClusterException("No way to dispatch this command to Redis Cluster. Missing key.")
        command = args[0]

        if command in ['EVAL', 'EVALSHA']:
            numkeys = args[2]
            keys = args[3: 3 + numkeys]
            slots = {self.connection_pool.nodes.keyslot(key) for key in keys}
            if len(slots) != 1:
                raise RedisClusterException("{0} - all keys must map to the same key slot".format(command))
            return slots.pop()
        elif command in ('XREAD', 'XREADGROUP'):
            try:
                idx = args.index('STREAMS') + 1
            except ValueError:
                raise RedisClusterException("{0} arguments do not contain STREAMS operand".format(command))
            key = args[idx]
        elif command in ('XGROUP', 'XINFO'):
            key = args[2]
        else:
            key = args[1]

        return self.connection_pool.nodes.keyslot(key)

    def _merge_result(self, command, res, **kwargs):
        """
        `res` is a dict with the following structure Dict(NodeName, CommandResult)
        """
        if command in self.result_callbacks:
            return self.result_callbacks[command](res, **kwargs)

        # Default way to handle result
        return first_key(res)

    def determine_node(self, *args, **kwargs):
        """
        TODO: document
        """
        command = args[0]
        node_flag = self.nodes_flags.get(command)

        if node_flag == NodeFlag.BLOCKED:
            return blocked_command(self, command)
        elif node_flag == NodeFlag.RANDOM:
            return [self.connection_pool.nodes.random_node()]
        elif node_flag == NodeFlag.ALL_MASTERS:
            return self.connection_pool.nodes.all_masters()
        elif node_flag == NodeFlag.ALL_NODES:
            return self.connection_pool.nodes.all_nodes()
        elif node_flag == NodeFlag.SLOT_ID:
            # if node flag of command is SLOT_ID
            # `slot_id` should is assumed in kwargs
            slot = kwargs.get('slot_id')
            if not slot:
                raise RedisClusterException('slot_id is needed to execute command {}'
                                            .format(command))
            return [self.connection_pool.nodes.node_from_slot(slot)]
        else:
            return None

    @clusterdown_wrapper
    async def execute_command(self, *args, **kwargs):
        """
        Sends a command to a node in the cluster
        """
        if not self.connection_pool.initialized:
            await self.connection_pool.initialize()
        if not args:
            raise RedisClusterException("Unable to determine command to use")

        command = args[0]

        node = self.determine_node(*args, **kwargs)
        if node:
            return await self.execute_command_on_nodes(node, *args, **kwargs)

        # If set externally we must update it before calling any commands
        if self.refresh_table_asap:
            await self.connection_pool.nodes.initialize()
            self.refresh_table_asap = False

        redirect_addr = None
        asking = False

        try_random_node = False
        slot = self._determine_slot(*args)
        ttl = int(self.RedisClusterRequestTTL)

        while ttl > 0:
            ttl -= 1

            if asking:
                node = self.connection_pool.nodes.nodes[redirect_addr]
                r = self.connection_pool.get_connection_by_node(node)
            elif try_random_node:
                r = self.connection_pool.get_random_connection()
                try_random_node = False
            else:
                if self.refresh_table_asap:
                    # MOVED
                    node = self.connection_pool.get_master_node_by_slot(slot)
                else:
                    node = self.connection_pool.get_node_by_slot(slot)
                r = self.connection_pool.get_connection_by_node(node)

            try:
                if asking:
                    await r.send_command('ASKING')
                    await self.parse_response(r, "ASKING", **kwargs)
                    asking = False

                await r.send_command(*args)
                return await self.parse_response(r, command, **kwargs)
            except (RedisClusterException, BusyLoadingError, CancelledError):
                raise
            except (ConnectionError, TimeoutError):
                try_random_node = True

                if ttl < self.RedisClusterRequestTTL / 2:
                    await asyncio.sleep(0.1)
            except ClusterDownError as e:
                self.connection_pool.disconnect()
                self.connection_pool.reset()
                self.refresh_table_asap = True

                raise e
            except MovedError as e:
                # Reinitialize on ever x number of MovedError.
                # This counter will increase faster when the same client object
                # is shared between multiple threads. To reduce the frequency you
                # can set the variable 'reinitialize_steps' in the constructor.
                self.refresh_table_asap = True
                await self.connection_pool.nodes.increment_reinitialize_counter()

                node = self.connection_pool.nodes.set_node(e.host, e.port, server_type='master')
                self.connection_pool.nodes.slots[e.slot_id][0] = node
            except TryAgainError as e:
                if ttl < self.RedisClusterRequestTTL / 2:
                    await asyncio.sleep(0.05)
            except AskError as e:
                redirect_addr, asking = "{0}:{1}".format(e.host, e.port), True
            finally:
                self.connection_pool.release(r)

        raise ClusterError('TTL exhausted.')

    async def execute_command_on_nodes(self, nodes, *args, **kwargs):
        command = args[0]
        res = {}

        for node in nodes:
            connection = self.connection_pool.get_connection_by_node(node)

            # copy from redis-py
            try:
                await connection.send_command(*args)
                res[node["name"]] = await self.parse_response(connection, command, **kwargs)
            except CancelledError:
                # do not retry when coroutine is cancelled
                connection.disconnect()
                raise
            except (ConnectionError, TimeoutError) as e:
                connection.disconnect()

                if not connection.retry_on_timeout and isinstance(e, TimeoutError):
                    raise

                await connection.send_command(*args)
                res[node["name"]] = await self.parse_response(connection, command, **kwargs)
            finally:
                self.connection_pool.release(connection)
        return self._merge_result(command, res, **kwargs)

    async def pipeline(self, transaction=None, shard_hint=None, watches=None):
        """
        Cluster impl:
            Pipelines do not work in cluster mode the same way they do in normal mode.
            Create a clone of this object so that simulating pipelines will work correctly.
            Each command will be called directly when used and when calling execute() will only return the result stack.
        cluster transaction can only be run with commands in the same node, otherwise error will be raised.
        """
        await self.connection_pool.initialize()
        if shard_hint:
            raise RedisClusterException("shard_hint is deprecated in cluster mode")

        from aredis.pipeline import StrictClusterPipeline
        return StrictClusterPipeline(
            connection_pool=self.connection_pool,
            startup_nodes=self.connection_pool.nodes.startup_nodes,
            result_callbacks=self.result_callbacks,
            response_callbacks=self.response_callbacks,
            transaction=transaction,
            watches=watches
        )
